docker/snaptools/bc_attacher.py

```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python
+
+# Author: Nick Barkas <nbarkas@broadinstitute.org>
+# Date: May 9th, 2019
+# Description: Pre-process fastq files by attaching barcodes to read names (generally a bad idea)
+
+import argparse
+from Bio import SeqIO
+import time
+import gzip
+
+parser = argparse.ArgumentParser(
+    description="Extract barcodes from index reads into read names for scATAC-seq BICCN data V1"
+)
+
+parser.add_argument(
+    '--index1',
+    required=True,
+    dest='file_i1',
+    help='index1 FASTQ file'
+)
+parser.add_argument(
+    '--index2',
+    required=True,
+    dest='file_i2',
+    help='index2 FASTQ file'
+)
+parser.add_argument(
+    '--read1',
+    required=True,
+    dest='file_r1',
+    help='read1 FASTQ file'
+)
+parser.add_argument(
+    '--read2',
+    required=True,
+    dest='file_r2',
+    help='read2 FASTQ files'
+)
+parser.add_argument(
+    '--out-read1',
+    required=True,
+    dest='out_file_r1',
+    help='output file read1'
+)
+parser.add_argument(
+    '--out-read2',
+    required=True,
+    dest='out_file_r2',
+    help='output file read2'
+)
+
+args = parser.parse_args()
+
+# Open the files
+fh_i1 = gzip.open(args.file_i1, 'rU')
+fh_i2 = gzip.open(args.file_i2, 'rU')
+fh_r1 = gzip.open(args.file_r1, 'rU')
+fh_r2 = gzip.open(args.file_r2, 'rU')
+# output files
+fh_out_r1 = gzip.open(args.out_file_r1, 'w')
+fh_out_r2 = gzip.open(args.out_file_r2, 'w')
+
+# Get iterators for the files
+iter_i1 = SeqIO.parse(fh_i1, 'fastq')
+iter_i2 = SeqIO.parse(fh_i2, 'fastq')
+iter_r1 = SeqIO.parse(fh_r1, 'fastq')
+iter_r2 = SeqIO.parse(fh_r2, 'fastq')
+
+start_time = time.time()
+
+nEntries = 0
+readComplete = False
+while not readComplete:
+    if (nEntries % 100000 == 0):
+        print("Processed %i records..." % nEntries)
+    try:
+        i1_sreq = iter_i1.next()
+        i2_sreq = iter_i2.next()
+        r1_sreq = iter_r1.next()
+        r2_sreq = iter_r2.next()
```

Comment:
The iterators could be stored in a list and the `map(lambda iter: iter.next(), iterators)` could be used to advance all at once.

Reply:
yes, also should add a check to make sure that the iterators reach the EOF at the same time. If not a warning should be issued to the user